### DISCUSSION TITLE
Verification of HTTP Requests fails ("Verification failed for @user...")

### DISCUSSION THREAD

**Discussion by sventiffe** [mastodon/mastodon#21621] [Q&A] [answered]
I want to send an Accept activity after receiving a Follow activity and am getting a 401 response with `Verification failed for bob@my-example.com https://my-example.com/actor using rsa-sha256 (RSASSA-PKCS1-v1_5 with SHA-256)` similar to what has been discussed [here](https://github.com/mastodon/mastodon/discussions/18453). 

This discussion helped me to fix some initial issues with my request. I've followed the documentation [here](https://docs.joinmastodon.org/spec/security/) and added a digest which seems to be required albeit not documented (issue [filed](https://github.com/mastodon/documentation/issues/1079)). As I still received a 401, I did set up a local development environment against which I am running my HTTP request to debug.

My request is failing at this [line](https://github.com/mastodon/mastodon/blob/3d906d8e62097d463a59b40c9498a47fdb850856/app/controllers/concerns/signature_verification.rb#L172): 
```
14:49:39 web.1     |     173: def verify_signature(actor, signature, compare_signed_string)
14:49:39 web.1     |  => 174:   if actor.keypair.public_key.verify(OpenSSL::Digest.new('SHA256'), signature, compare_signed_string)
14:49:39 web.1     |     175:     @signed_request_actor = actor
14:49:39 web.1     |     176:     @signed_request_actor
14:49:39 web.1     |     177:   end
14:49:39 web.1     |     178: rescue OpenSSL::PKey::RSAError
14:49:39 web.1     |     179:   nil
14:49:39 web.1     |     180: end
```

For context, `compare_signed_string` is the string that is being constructed on server side based on header data and includes the four headers `(request-target), host, date, digest)`. I’ve verified that `compare_signed_string` is identical with the string used on client side to compute the hash and signature. The strings are not only identical but also the computed hashs are identical on server side:
```
14:53:11 web.1     | [47] pry(#<ActivityPub::InboxesController>)> OpenSSL::Digest.new('sha256',compare_signed_string).to_s
14:53:11 web.1     | => "2fe8330cac4ceae644bb035909b42225c228605dbefd92e410601b4f99d5480f"
``` 
and client side:
```
$ od -t x1 signed_string.sha256
2f  e8  33  0c  ac  4c  ea  e6  44  bb  03  59  09  b4  22  25
c2  28  60  5d  be  fd  92  e4  10  60  1b  4f  99  d5  48  0f
```

In a similar way I’ve compared the signature on the server side (`p signature.to_json_raw` during runtime) and client side (`od -tu1 signed_string.sha256.sig` on an exported file). 

I’ve also double checked that the object actor on server side has the correct public key: 
```
p actor
14:53:33 web.1     |  public_key:
14:53:33 web.1     |   "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxrp/NfcTEaiSl0pyOFuP\nbbO3ym6MXfSAUjb9Lg1J6FIW6wjbL8E9C2r4J5AEOVXI6KoLb8Nx2puegItp4nF9\nPCnil1coK7clAOxTKCZbkyupJ47DdLqJTIc04JBVkhDKnYHIclN/ZsUMFGzHX2vG\nKaF+SZyagv7B+qV/w3wLUrUMfRtPcnWV7oG1mngAEnAjp63dAR7VkyK0IqPYDy/f\nWI2SGbEOfMSPWcxgOVQRG0Du4pD4iWKkEx38u1Os4iV4VgivsO/JUirLvvfiZjsH\nlYvfLEPnklVr/nOMHmkShpPKqFS4/15SvuwNVbIB+GRefbHKkHXmWJAesh2PW5yW\nEwIDAQAB\n-----END PUBLIC KEY-----",
```

Using this key on client side as pem file, I can verify the signature on my client as follows:
```
$ openssl rsautl -verify -in signed_string.sha256.sig -inkey public.pem -pubin > hash.bytes
```

The extracted hash during this verification step on the client is identic with the computed (ie expected hash) on the server side:
```
$ od -t x1 hash.bytes 
2f  e8  33  0c  ac  4c  ea  e6  44  bb  03  59  09  b4  22  25
c2  28  60  5d  be  fd  92  e4  10  60  1b  4f  99  d5  48  0f
```

So despite the validity of data on client and server side, and despite my ability to verify the signature on client side, still the verification on server side at this point fails, as this statement returns false: `actor.keypair.public_key.verify(OpenSSL::Digest.new('SHA256'), signature, compare_signed_string)`

I’ve only started to look into this recently and I’ve had my first Ruby debugging attempt ever today, so I might miss something completely obvious. I have no idea what and how to further investigate. Any help appreciated!

Fwiw, these are the Java code snippets for creating the hash and signing it:
```
// Hash the signature string with SHA256
byte[] messageHash = MessageDigest.getInstance("SHA-256")
    .digest(signatureString.getBytes());

// Compute signature from private key that is read as PEM file
KeyFactory keyFactory = KeyFactory.getInstance("RSA");
RSAPrivateKey privateKey = (RSAPrivateKey)keyFactory.generatePrivate(
    new PKCS8EncodedKeySpec(Base64.getDecoder().decode(privateKeyPEM)));

Cipher cipher = Cipher.getInstance("RSA");
cipher.init(Cipher.ENCRYPT_MODE, privateKey);
byte[] digitalSignature = cipher.doFinal(messageHash);
```

**Comment by ClearlyClaire** [comment]
Are you able to run your own Mastodon server? If so, you may want to merge #21617 and try to see if you can spot your issue with the improved debugging feedback

**Reply by sventiffe** [reply]
Thank you! The error message on client end is better now but does not give me any additional hint:
```
{
   "error":"Verification failed for bob@my-example.com https://my-example.com/actor using rsa-sha256 (RSASSA-PKCS1-v1_5 with SHA-256)",
   "signed_string":"(request-target): post /inbox\nhost: 192.168.1.101:3000\ndate: Fri, 25 Nov 2022 18:04:45 GMT\ndigest: SHA-256=ILMNSQW3KLUz2BuSp3Vtma7/+XNuyik6VGmlxhJGdFk=",
   "signature":"NVS9k9xbWFK7W8LhrfXEdS6syFiT1TpbU9FjHXYQJPL0sEfyCd0W77A+yj649H0/+1AAWMIvNYtA6bEMkvE89Rl8i6Pd6f/I6+Fkkq5lQiuM475SgNu99HpVv45UeHViP/vp+X85MD5fqcQG3uGtD8FWB3H3kP0gBCKDr6dAV+SxEHYZnKFy4HC8/tg1B78/xPHlsWhVcmAVGR/Fm6rS7D/dJJ86SD4+KIKFi43MVUskm99ssdyLRsRfcBBABz09HmcUiuBWyXi5VuzPWflFxlVyA0OuA11XbL2dJj7bsSrgu14a0le2GBe8zDwz+/V5WIr1XNdHTUBgXyXQTfGuvQ=="
}
```
I could not see any particular issues in the log as well.

Is there a way to debug why `actor.keypair.public_key.verify(OpenSSL::Digest.new('SHA256'), signature, compare_signed_string)` returns false although the verification of the signature on command line actually returns the correct hash for the string? 

If it helps, this is the public key that is being used for the verification:
```
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxrp/NfcTEaiSl0pyOFuP
bbO3ym6MXfSAUjb9Lg1J6FIW6wjbL8E9C2r4J5AEOVXI6KoLb8Nx2puegItp4nF9
PCnil1coK7clAOxTKCZbkyupJ47DdLqJTIc04JBVkhDKnYHIclN/ZsUMFGzHX2vG
KaF+SZyagv7B+qV/w3wLUrUMfRtPcnWV7oG1mngAEnAjp63dAR7VkyK0IqPYDy/f
WI2SGbEOfMSPWcxgOVQRG0Du4pD4iWKkEx38u1Os4iV4VgivsO/JUirLvvfiZjsH
lYvfLEPnklVr/nOMHmkShpPKqFS4/15SvuwNVbIB+GRefbHKkHXmWJAesh2PW5yW
EwIDAQAB
-----END PUBLIC KEY-----
```

**Reply by ClearlyClaire** [reply]
No, it's not possible to provide more granular feedback without the private key (which the server obviously does not have). I'd first check that the `signed_string` is indeed character-by-character identical to what you're signing.

**Reply by ClearlyClaire** [reply]
Also I think you may want to use https://docs.oracle.com/javase/7/docs/api/java/security/Signature.html

**Reply by ClearlyClaire** [reply]
Sorry, I had skimmed too fast while reading the issue and missed you verified `signed_string` was what you intended. I think this may be down to RSASSA-PKCS1-v1_5 using padding and not your code, but this is just a guess.

**Reply by sventiffe** [reply]
🎉 202 - Accepted 🎉

```
Signature signer = Signature.getInstance("SHA256WithRSA");
signer.initSign(privateKey);
signer.update(originalString.getBytes());
byte[] digitalSignature = signer.sign();
```

So what brought me on the wrong track was to literally follow the [instruction](https://docs.joinmastodon.org/spec/security/):

> The signature string is then hashed with SHA256 and signed with the actor’s public key.

Being a n00b in this area I did not consider anything else (I should have noticed the "RSASSA-PKCS1-v1_5" part in the error message though which might have brought me on the right track). I've updated the documentation bug that I've filed and I might send a PR for it later.

**Comment by nightpool** [comment]
Considering that you've already confirmed `compare_signed_string` is identical to your value and that the signature validates when using OpenSSL locally, could it be possible that the value of actor.public_key is incorrect? Maybe it got updated at some point and the Mastodon server is still using the old version? 

Also, this part was slightly confusing to me

> In a similar way I’ve compared the signature on the server side (p signature.to_json_raw during runtime) and client side (od -tu1 signed_string.sha256.sig on an exported file).

One of those is operating on hex encoded values and one is operating on raw bytes, correct? Are you passing the digest hex encoded, or as raw bytes?

**Reply by ClearlyClaire** [reply]
It refreshes keys on verification failures, and fail with an explicit message if refreshing fails. But maybe the place where it fetches it contains outdated info.

**Reply by sventiffe** [reply]
> One of those is operating on hex encoded values and one is operating on raw bytes, correct? 

Both are byte data; the variable `signature` at the break point on server side looks like a byte array and poking around I found `.to_json_raw` for "prettifying" it[1]. ` od -tu1` on command line also converts the client-side signature which I've stored to disk to unsigned byte integers, so I've compared the output of both byte-by-byte.

> could it be possible that the value of actor.public_key is incorrect

The public key on server side which I've inspected through `p actor` on the breakpoint is identical with the content of the PEM file on disk used for local validation. It just has `\n` inline for the line breaks (which I assume are removed on parsing it anyway).

[1] as I've mentioned, this has been my very first contact with Ruby, so maybe i did something quite stupid, but the output looked consistent

**Comment by sventiffe** [comment, marked as answer]
Summarizing the answer from the comments, thanks @ClearlyClaire!

> I think this may be down to RSASSA-PKCS1-v1_5 using padding and not your code, but this is just a guess.

> Also I think you may want to use https://docs.oracle.com/javase/7/docs/api/java/security/Signature.html

Signing the string with this code solved the issue:
```
Signature signer = Signature.getInstance("SHA256WithRSA");
signer.initSign(privateKey);
signer.update(originalString.getBytes());
byte[] digitalSignature = signer.sign();
```
